### PR TITLE
Add image size validation

### LINKED ISSUE #1385: Image size validator
There exists a format validator, so you couldn't upload files with an unsupported file type.

There should also be a validator, so you couldn't upload an image with 10x10 px where the minimum image size is 800x600 px.

### PR DESCRIPTION
**Pull request by core23**
## Subject

I am targeting this branch, because this feature respects BC.

Closes #1385

## Changelog

```markdown
### Added
- Add image size validation
```

### CODE REVIEW

File: src/Provider/ImageProvider.php [hunk 1]
```diff
@@ -207,6 +208,49 @@ public function generatePrivateUrl(MediaInterface $media, $format)
...
+        }
+
+        $size = $image->getSize();
+
+        if ($size->getWidth() < $minWidth || $size->getHeight() < $minHeight) {
+            $errorElement
+                ->with('binaryContent')
+                    ->addViolation('error.image_too_small', [
+                        '%min%' => $minWidth,
+                        '%max%' => $minHeight,
```
Comment (phansys) [inline src/Provider/ImageProvider.php:248]: I guess the validation against max sizes is still in progress.
```suggestion
                        '%min_width%' => $minWidth,
                        '%min_height%' => $minHeight,
```

**Comment by wbloszyk** [comment]
Can you add max too? This will be nice feature for non resize provider.

**Comment by core23** [comment]
> Can you add max too? This will be nice feature for non resize provider.

This is an other feature to me. The first goal is to add a useful constraint, if you want to resize the source image to a larger size.

File: src/Provider/ImageProvider.php [hunk 2]
```diff
@@ -15,6 +15,7 @@
 
 use Gaufrette\Filesystem;
 use Imagine\Image\ImagineInterface;
+use Sonata\CoreBundle\Validator\ErrorElement;
```
Comment (wbloszyk) [inline src/Provider/ImageProvider.php:18]:
```suggestion
use Sonata\Form\Validator\ErrorElement;
```

File: src/Provider/ImageProvider.php [hunk 3]
```diff
@@ -243,4 +287,20 @@ protected function doTransform(MediaInterface $media)
...
+    {
+        $formats = [];
+
+        foreach ($this->getFormats() as $providerFormat => $settings) {
+            if (0 === strpos($providerFormat, $context)) {
+                $formats[$providerFormat] = $settings;
+            }
+        }
+
+        return $formats;
```
Comment (phansys) [inline src/Provider/ImageProvider.php:304]:
```suggestion
        return array_filter($this->getFormats(), static function (string $providerFormat) use ($context): bool {
            return 0 === strpos($providerFormat, $context);
        }, ARRAY_FILTER_USE_KEY);
```

File: src/Resources/translations/SonataMediaBundle.en.xliff [hunk 4]
```diff
@@ -571,6 +571,10 @@
                 <source>form.label_icon</source>
                 <target>Icon</target>
             </trans-unit>
+            <trans-unit id="error.image_too_small">
+                <source>error.image_too_small</source>
+                <target>Image too small, min size %min_width% x %min_height%</target>
```
Comment (phansys) [inline src/Resources/translations/SonataMediaBundle.en.xliff:576]:
```suggestion
                <target>Image too small, it requires %min_width% x %min_height% as minimum.</target>
```

File: src/Provider/ImageProvider.php [hunk 5]
```diff
@@ -207,6 +208,49 @@ public function generatePrivateUrl(MediaInterface $media, $format)
         return $this->thumbnail->generatePrivateUrl($this, $media, $format);
     }
 
+    public function validate(ErrorElement $errorElement, MediaInterface $media)
```
Comment (greg0ire) [inline src/Provider/ImageProvider.php:211]:
```suggestion
    public function validate(ErrorElement $errorElement, MediaInterface $media): void
```

File: src/Provider/ImageProvider.php [hunk 6]
```diff
@@ -207,6 +208,49 @@ public function generatePrivateUrl(MediaInterface $media, $format)
...
+        }
+
+        if (0 === $minWidth && 0 === $minHeight) {
+            return;
+        }
+
+        try {
+            $image = $this->imagineAdapter->open($media->getBinaryContent()->getPathname());
+        } catch (\RuntimeException $e) {
+            return;
```
Comment (greg0ire) [inline src/Provider/ImageProvider.php:238]: Shouldn't we log something here? Or at least add a comment?

File: src/Provider/ImageProvider.php (hunk 5 above)
Comment (core23) [reply src/Provider/ImageProvider.php:211]: The class implements an interface with not signature, so we should not use return types (yet)

File: src/Provider/ImageProvider.php (hunk 6 above)
Comment (core23) [reply src/Provider/ImageProvider.php:238]: It's a copy & paste of https://github.com/sonata-project/SonataMediaBundle/blob/8dd4990922efbc4163f528b77a6a677f3027f1bc/src/Provider/ImageProvider.php#L235-L239

If we can an error, we would trigger a validation error in an other component.

**Comment by wbloszyk** [comment]
Can you move this constraint to seperate files and allow register multiple constraints for each provider/formats? Like in form field? This will allow add max filesize and other in feature.

File: src/Provider/ImageProvider.php (hunk 6 above)
Comment (greg0ire) [reply src/Provider/ImageProvider.php:238]: > If we can an error, we would trigger a validation error in an other component.

Sorry, can you fix that sentence please? I don't understand it.

Comment (core23) [reply src/Provider/ImageProvider.php:239]: Added a comment in the code

File: tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+
+        return $image;
+    }
+
+    /**
+     * @return MockObject|MediaInterface
+     */
+    protected function mockMedia()
+    {
+        $binaryContent = $this->createMock(UploadedFile::class);
```
Comment (phansys) [inline tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php:148]:
```suggestion
        $binaryContent = $this->createStub(UploadedFile::class);
```

File: tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        $binaryContent->method('getPathname')->willReturn(tmpfile());
+
+        $media = $this->createMock(MediaInterface::class);
```
Comment (phansys) [inline tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php:151]:
```suggestion
        $media = $this->createStub(MediaInterface::class);
```

File: tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+            $this->imageProvider
+        );
+    }
+
+    /**
+     * @return ImageInterface|MockObject
+     */
+    protected function mockImage(int $width, int $height)
+    {
+        $box = $this->createMock(BoxInterface::class);
```
Comment (phansys) [inline tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php:133]:
```suggestion
        $box = $this->createStub(BoxInterface::class);
```

File: tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+        $box->method('getWidth')->willReturn($width);
+        $box->method('getHeight')->willReturn($height);
+
+        $image = $this->createMock(ImageInterface::class);
```
Comment (phansys) [inline tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php:137]:
```suggestion
        $image = $this->createStub(ImageInterface::class);
```

File: tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+    private $imagineAdapter;
+
+    /**
+     * @var ImageProvider
+     */
+    private $imageProvider;
+
+    protected function setUp(): void
+    {
+        $this->imagineAdapter = $this->createMock(ImagineInterface::class);
```
Comment (phansys) [inline tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php:46]:
```suggestion
        $this->imagineAdapter = $this->createStub(ImagineInterface::class);
```

File: tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+        $this->imageProvider = $this->createMock(ImageProvider::class);
```
Comment (phansys) [inline tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php:47]:
```suggestion
        $this->imageProvider = $this->createStub(ImageProvider::class);
```

File: src/Validator/Constraints/ImageUploadDimension.php [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace Sonata\MediaBundle\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+
+final class ImageUploadDimension extends Constraint
+{
+    public $message = 'error.image_too_small';
+
+    public function getTargets()
```
Comment (greg0ire) [inline src/Validator/Constraints/ImageUploadDimension.php:22]:
```suggestion
    public function getTargets(): string
```

File: tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+        return new ImageUploadDimensionValidator(
+            $this->imagineAdapter,
+            $this->imageProvider
+        );
+    }
+
+    /**
+     * @return ImageInterface|MockObject
+     */
+    protected function mockImage(int $width, int $height)
```
Comment (greg0ire) [inline tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php:131]:
```suggestion
    private function mockImage(int $width, int $height): ImageInterface
```

File: tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+    protected function createValidator(): ConstraintValidatorInterface
+    {
+        return new ImageUploadDimensionValidator(
+            $this->imagineAdapter,
+            $this->imageProvider
+        );
+    }
+
+    /**
+     * @return ImageInterface|MockObject
```
Comment (greg0ire) [inline tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php:129]:
```suggestion
     * @return ImageInterface&MockObject
```

File: tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+        $box->method('getHeight')->willReturn($height);
+
+        $image = $this->createStub(ImageInterface::class);
+        $image->method('getSize')->willReturn($box);
+
+        return $image;
+    }
+
+    /**
+     * @return MockObject|MediaInterface
```
Comment (greg0ire) [inline tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php:144]:
```suggestion
     * @return MockObject&MediaInterface
```

File: tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php [hunk 17]
```diff
@@ -0,0 +1,157 @@
...
+     */
+    protected function mockMedia()
```
Comment (greg0ire) [inline tests/Fixtures/SonataClassificationBundle/Validator/Constraints/ImageUploadDimensionValidatorTest.php:146]:
```suggestion
    private function mockMedia(): MediaInterface
```

**Pull request merged**